tests/test_completion.py

```diff
@@ -53,6 +54,49 @@ def create_config_loader() -> ConfigLoaderImpl:
     )
 
 
+@pytest.mark.skipif(sys.platform == "win32", reason="does not run on windows")
+class TestDotInPathCompletion:
+    def test_bash_completion_with_dot_in_path(self) -> None:
+        process = subprocess.Popen(
+            textwrap.dedent(
+                """
+                bash <<- 'HEREDOC'
+                # Navigate to app with minimal configuration
+                cd tests/test_apps/app_with_cfg/
+
+                # Locate python
+                PYTHON=$(command -v python)
+
+                # Setup hydra bash completion
+                eval "$(python my_app.py -sc install=bash)"
+
+                # Setup debugging
+                export HYDRA_COMP_DEBUG=1
+                export PATH=$PATH:.
+
+                # Do actual test
+                test=$(COMP_LINE='python my_app.py' hydra_bash_completion |\
+                    grep EXECUTABLE_FIRST |\
+                    awk '{split($0,a,"=");b=substr(a[2],2,length(a[2])-2);print b}')
+
+                if [ $test == $PYTHON ]; then
+                    echo TRUE
+                else
+                    echo FALSE
+                fi
+                HEREDOC
+                """
+            ),
+            shell=True,
```

Comment:
`shell=True` is insecure in general.
Can you switch to keeping the shell script in a simple file and calling with subprocess.check_output() ?

Reply:
This is not difficult, but against my habits. In the projects I worked so far we use pytest (and similar things for C/C++/Fortran) and we always make our tests as self contained as possible. So we keep all things in one place - to give ultimate visibility of what is really being tested at particular test.
I would actually suggest you to change some of your expect scripts to my format to make sure that everything in one place.
I have an anecdotal story when push from my colleague changed many files at ones and one of them was part of the test suite - but then at final merge another colleague changed how that test was executed (unintentionally) which caused test to pass for all cases (even where it should have been failed).
If that test would be in the script (like the one above) - this would not happen.

P.S. I have to admit - I never used mypy and nox before - and given what I've seen I will probably not use them any time soon - IMHO - until they will mature a bit more and will make things more automatic.

Reply:
Okay, I accept your reasoning. but do get rid of the shell=True somehow.

As for mypy and nox:
1. nox is fantastic. What issues did you have with it?
2. mypy is industry standard.
The CI is intentionally testing a lot of things. It does take time to handle all the requirements, as especially given that it's your first PR.

Reply:
As for the expect script: I am expecting to have a change to it every 10 years.

Reply:
This is your repo - you are the boss here - I'll split it into separate file and will make shell=False.
I said - it's easy - was just sharing some of my personal observations.
P.S. I just take testing easy - nox is like "heavy metal for relaxation" for me (don't take me wrong - I love heavy metal - but consider it to be relaxing only during Armageddons). This is indeed my first PR with mypy - I guess I don't belong to the same industry :-)).

Reply:
I agree with your reasoning to keep it in one place.
what I am suggesting is that your test can dump a temp file from the script and execute it with bash.

Reply:
I am being strict on testing because I hate breaking things unknowingly.
If something breaks, it would better be because I intentionally broke it and not because of poor test coverage.

Reply:
Creating files on the fly for the tests is OK if it's data files - but executable scripts - that's something which I try to avoid due to the way different systems handle temporary files and setting permissions - so I decided to split it all together.

Reply:
1. sure.
2. This is not an executable, just a text file that is being executed by bash.
3. This shows you the power of good CI. if it your tests pass in CI you know it works on Linux Mac and Windows.